README.md
# GreenCape Extension Export

This is a **system plugin** for Joomla **3.8+** that extends Joomla's Extension Manager with extension export functionality.

## Installation

Download the latest release (see [Release tab](https://github.com/GreenCape/extension-export/releases)) and install it like any other plugin.
Navigate to *Extensions/Plugins*, and search for `extension export`.
Enable the plugin.

## Configuration

When you click on the plugin's name in the view mentioned before, you can edit the settings.

### Directory

Specify the directory, where the exported packages will be stored for later use.
It defaults to `images/dist`, so the packages can easily be managed (renamed, deleted) using the Media Manager.

### Delete Directories

This switch controls the deletion of the temporary directories after bundling the extension files.
* `Yes` - delete the directories (default). Use this setting if you do not need the uncompressed packages. They can only be deleted if you have file system access (FTP or SSH).
* `No` - keep the directories. Use this setting (in your development environment) to simply copy the packages into your development area without unpacking them first.

### File Mode

The File Mode is the UNIX permissions for files, defaulting to `0644`.

### Dir Mode

The Dir Mode is the UNIX permissions for directories, defaulting to `0755`.
This field is only available, if `Delete Directories` is set to `no`.

## Usage

Navigate to *Extensions/Manage/Manage*.
In the toolbar, a new button has appeared.

![The new "Export" button in the Manage Extensions view](docs/screenshot-button.png)

Select the extension that you want to export.
Currently, components, modules, plugins and templates are supported.
Click on the *Export* button.
If everything goes right, you'll see a success message with a download link.

![The success message and the download link](docs/screenshot-download.png)



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

To update from 1.0.0, please uninstall the old version and install the new one.

### Added

- Ability to download all selected extensions as a single package.
- Ability to keep unzipped directories.
- Support for Joomla Update System
- Added German language files

### Changed

- Default permissions to `0644` (files) / `0755` (directories). Thanks @brianteeman

### Deprecated

### Removed

### Fixed

### Security

## [1.0.0](https://github.com/GreenCape/extension-export/tree/1.0.0) - 2018-01-04

Initial Release.


src/lib/Zipper.php
<?php
/**
 * @package     GreenCape Extension Exporter
 * @author      Niels Braczek <nbraczek@bsds.de>
 *
 * @copyright   Copyright (C) 2012 - 2017 BSDS Braczek Software- und DatenSysteme. All rights reserved.
 * @license     GNU General Public License version 3 or later; see LICENSE.txt
 */

namespace GreenCape\Extension;

use /** @noinspection PhpDeprecationInspection */
	JArchiveZip;

defined('_JEXEC') or die;

/**
 * ExtensionExport Zipper
 *
 * @since  __DEPLOY_VERSION__
 */
class Zipper
{
	/**
	 * Files to include in the archive
	 *
	 * @var string[][]
	 *
	 * @since  __DEPLOY_VERSION__
	 */
	private $files = [];

	/**
	 * Add a file to the archive
	 *
	 * @param string $name Name of the file
	 * @param string $data Content of the file
	 * @param int    $time UNIX timestamp of creation time
	 *
	 * @since __DEPLOY_VERSION__
	 */
	public function addFile($name, $data, $time)
	{
		$this->files[] = [
			'name' => $name,
			'data' => $data,
			'time' => $time,
		];
	}

	/**
	 * Create the archive
	 *
	 * @param string $archive Filename of the archive
	 *
	 * @since __DEPLOY_VERSION__
	 */
	public function create($archive)
	{
		/** @noinspection PhpDeprecationInspection No replacement available */
		(new JArchiveZip)->create($archive, $this->files);
	}
}


src/lib/Packager.php
<?php
/**
 * @package     GreenCape Extension Exporter
 * @author      Niels Braczek <nbraczek@bsds.de>
 *
 * @copyright   Copyright (C) 2012 - 2017 BSDS Braczek Software- und DatenSysteme. All rights reserved.
 * @license     GNU General Public License version 3 or later; see LICENSE.txt
 */

namespace GreenCape\Extension;

use JPath;

defined('_JEXEC') or die;

/**
 * ExtensionExport Packager
 *
 * @since  __DEPLOY_VERSION__
 */
class Packager
{
	/**
	 * Data for use in the package manifest
	 *
	 * @var string[]
	 *
	 * @since  __DEPLOY_VERSION__
	 */
	protected $data = [
		'name'         => '',
		'author'       => '',
		'creationDate' => '',
		'packagename'  => 'exported',
		'version'      => '0.0.0',
		'url'          => '',
		'packager'     => 'GreenCape Extension Exporter',
		'packagerurl'  => 'https://github.com/GreenCape/extension-export',
		'copyright'    => '',
		'description'  => 'This package was exported using GreenCape Extension Exporter.',
	];

	/**
	 * Data for the files section of the package manifest
	 *
	 * @var string[][]
	 *
	 * @since  __DEPLOY_VERSION__
	 */
	private $files = [];

	/**
	 * Create the package
	 *
	 * @param array  $packageData
	 * @param string $exportPath
	 *
	 * @return string
	 * @throws \UnexpectedValueException
	 *
	 * @since  __DEPLOY_VERSION__
	 */
	public function export($packageData, $exportPath)
	{
		$authors    = [];
		$copyrights = [];
		$version    = '0.0.0';

		$zipper = new Zipper();

		foreach ($packageData as $data)
		{
			$authors[]    = $data->author;
			$copyrights[] = $data->copyright;
			$version      = version_compare($version, $data->version, '<') ? $data->version : $version;
			$this->addFile($data->filename, $data->type, $data->element, $data->folder);
			$zipper->addFile(
				$data->filename,
				file_get_contents($exportPath . '/' . $data->filename),
				filemtime($exportPath . '/' . $data->filename)
			);
		}

		$this->data['name']      = ucfirst($this->data['packagename']) . ' Extension Package';
		$this->data['author']    = $this->combineAuthors($authors);
		$this->data['copyright'] = $this->combineCopyrights($copyrights);
		$this->data['version']   = $version;

		$element = "pkg_{$this->data['packagename']}";

		$zipper->addFile(
			"{$element}.xml",
			$this->nodeToString($this->getManifest()),
			time()
		);

		$zipper->create(JPath::clean("{$exportPath}/{$element}-{$version}.zip"));

		return "$element-$version";
	}

	/**
	 * Set the package name
	 *
	 * @param string $packageName
	 *
	 * @return Packager
	 *
	 * @since  __DEPLOY_VERSION__
	 */
	public function setPackageName($packageName)
	{
		$this->data['packagename'] = $packageName;

		return $this;
	}

	/**
	 * Combine author names from the included extensions
	 *
	 * The unique authors are combined forming a comma separated list.
	 *
	 * @param array $authors
	 *
	 * @return string
	 *
	 * @since __DEPLOY_VERSION__
	 */
	private function combineAuthors($authors)
	{
		return implode(', ', array_unique($authors));
	}

	/**
	 * Combine copyright notices from the included extensions
	 *
	 * The substrings 'Copyright (C)' and 'All rights reserved.' are removed from all entries, if present.
	 * The remaining unique entries are combined forming a comma separated list.
	 * The substrings are added again, if they were present in at least one extension.
	 *
	 * @param string[] $copyrights
	 *
	 * @return string
	 *
	 * @since  __DEPLOY_VERSION__
	 */
	private function combineCopyrights($copyrights)
	{
		$foundCopyright = false;
		$foundReserved  = false;
		foreach ($copyrights as $key => $copyright)
		{
			if (preg_match('~^\s*(Copy\w+\s*)?(\(C\)\s*)?~i', $copyright, $match))
			{
				$copyright      = str_replace($match[0], '', $copyright);
				$foundCopyright = trim($match[0] > '');
			}
			if (preg_match('~\s*\.?\s*All rights reserved\W*$~', $copyright, $match))
			{
				$copyright     = str_replace($match[0], '', $copyright);
				$foundReserved = true;
			}
			$copyrights[$key] = $copyright;
		}

		return ($foundCopyright ? 'Copyright (C) ' : '') . trim(implode(', ',
				array_unique($copyrights))) . ($foundReserved ? '. All rights reserved.' : '');
	}

	/**
	 * Create the manifest from the collected data
	 *
	 * @return \SimpleXMLElement
	 *
	 * @since  __DEPLOY_VERSION__
	 */
	private function getManifest()
	{
		$this->data['creationDate'] = date('Y-m-d');

		$manifest = simplexml_load_string('<extension version="3.3.0" type="package" method="upgrade"></extension>');
		foreach ($this->data as $tag => $content)
		{
			if (empty($content))
			{
				continue;
			}

			$manifest->addChild($tag, $content);
		}

		$files = $manifest->addChild('files');

		foreach ($this->files as $filename => $attributes)
		{
			$node = $files->addChild('file', $filename);

			foreach ($attributes as $attribute => $value)
			{
				if (empty($value))
				{
					continue;
				}

				$node->addAttribute($attribute, $value);
			}
		}

		return $manifest;
	}

	/**
	 * Add a file to the manifest
	 *
	 * @param string $filename
	 * @param string $type
	 * @param string $id
	 * @param string $group
	 *
	 * @return Packager
	 *
	 * @since  __DEPLOY_VERSION__
	 */
	private function addFile($filename, $type, $id, $group = null)
	{
		$this->files[$filename] = compact('type', 'id', 'group');

		return $this;
	}

	/**
	 * Convert a SimpleXMLElement to a pretty printed XML string
	 *
	 * @param \SimpleXMLElement $node
	 *
	 * @return string
	 *
	 * @since __DEPLOY_VERSION__
	 */
	private function nodeToString($node)
	{
		$dom = new \DOMDocument('1.0', 'utf-8');
		$dom->appendChild($dom->importNode(dom_import_simplexml($node), true));
		$dom->formatOutput = true;

		return $dom->saveXML();
	}
}


src/lib/DataMapper.php
<?php
/**
 * @package     GreenCape Extension Exporter
 * @author      Niels Braczek <nbraczek@bsds.de>
 *
 * @copyright   Copyright (C) 2012 - 2017 BSDS Braczek Software- und DatenSysteme. All rights reserved.
 * @license     GNU General Public License version 3 or later; see LICENSE.txt
 */

namespace GreenCape\Extension;

use Joomla\CMS\Factory;

defined('_JEXEC') or die;

/**
 * ExtensionExport DataMapper
 *
 * @since  1.0.0
 */
class DataMapper
{
	/**
	 * Get extension details from database
	 *
	 * @param int $extensionId
	 *
	 * @return \stdClass
	 * @throws \RuntimeException
	 *
	 * @since 1.0.0
	 */
	public function getExtensionDetails($extensionId)
	{
		$db    = Factory::getDbo();
		$query = $db->getQuery(true);
		$query
			->select('package_id, element, type, client_id, folder, manifest_cache')
			->from('#__extensions')
			->where("extension_id = $extensionId")
		;
		$db->setQuery($query);

		$attributes    = $db->loadAssoc();
		$manifestCache = json_decode($attributes['manifest_cache'], true);

		return (object) array_merge($attributes, $manifestCache);
	}
}


src/extensionexport.php
<?php
/**
 * @package     GreenCape Extension Exporter
 * @author      Niels Braczek <nbraczek@bsds.de>
 *
 * @copyright   Copyright (C) 2012 - 2017 BSDS Braczek Software- und DatenSysteme. All rights reserved.
 * @license     GNU General Public License version 3 or later; see LICENSE.txt
 */

use Joomla\CMS\Plugin\CMSPlugin;

defined('_JEXEC') or die;

/**
 * ExtensionExport Plugin.
 *
 * @since  1.0.0
 */
class PlgSystemExtensionExport extends CMSPlugin
{
	/** @noinspection GenericObjectTypeUsageInspection */
	/**
	 * Constructor.
	 *
	 * @param   object $subject The object to observe.
	 * @param   array  $config  An optional associative array of configuration settings.
	 *
	 * @since   1.0.0
	 * @throws \RuntimeException
	 */
	public function __construct($subject, $config)
	{
		$this->autoloadLanguage = true;

		parent::__construct($subject, $config);

		JLoader::register(
			'InstallerControllerExtensions',
			__DIR__ . '/lib/InstallerControllerExtensions.php'
		);
		JLoader::register(
			'InstallerViewDefault',
			JPATH_ADMINISTRATOR . '/components/com_installer/views/default/view.php'
		);
		JLoader::register(
			'InstallerViewManage',
			__DIR__ . '/lib/InstallerViewManage.php'
		);
		JLoader::register(
			'ContentHelper',
			JPATH_ADMINISTRATOR . '/components/com_content/helpers/content.php'
		);
		JLoader::registerNamespace(
			'GreenCape\\Extension\\',
			__DIR__ . '/lib',
			false,
			false,
			'psr4'
		);
	}
}


src/lib/InstallerViewManage.php
<?php
/**
 * @package     GreenCape Extension Exporter
 * @author      Niels Braczek <nbraczek@bsds.de>
 *
 * @copyright   Copyright (C) 2012 - 2017 BSDS Braczek Software- und DatenSysteme. All rights reserved.
 * @license     GNU General Public License version 3 or later; see LICENSE.txt
 */

use Joomla\CMS\Form\Form;
use Joomla\CMS\HTML\HTMLHelper;
use Joomla\CMS\Pagination\Pagination;
use Joomla\Registry\Registry;

defined('_JEXEC') or die;

/** @noinspection PhpUndefinedClassInspection */

/**
 * Extension Manager Manage View
 *
 * @since  1.0.0
 */
class InstallerViewManage extends InstallerViewDefault
{
	/**
	 * The selected extensions
	 *
	 * @var array
	 * @since 1.0.0
	 */
	protected $items;

	/**
	 * The pagination
	 *
	 * @var Pagination
	 * @since 1.0.0
	 */
	protected $pagination;

	/**
	 * The form
	 *
	 * @var Form
	 * @since 1.0.0
	 */
	protected $form;

	/**
	 * The current state
	 *
	 * @var Registry
	 * @since 1.0.0
	 */
	protected $state;

	/**
	 * The filter form
	 *
	 * @var Form
	 * @since __DEPLOY_VERSION__
	 */
	public $filterForm;

	/**
	 * The active filters
	 *
	 * @var
	 * @since __DEPLOY_VERSION__
	 */
	private $activeFilters;

	/**
	 * Display the view.
	 *
	 * @param   string $tpl Template
	 *
	 * @return  void
	 * @throws RuntimeException
	 *
	 * @since   1.0.0
	 */
	public function display($tpl = null)
	{
		$this->state         = $this->get('State');
		$this->items         = $this->get('Items');
		$this->pagination    = $this->get('Pagination');
		$this->filterForm    = $this->get('FilterForm');
		$this->activeFilters = $this->get('ActiveFilters');

		if (count($errors = $this->get('Errors')))
		{
			throw new RuntimeException(implode("\n", $errors), 500);
		}

		HTMLHelper::addIncludePath(JPATH_ADMINISTRATOR . '/components/com_installer/helpers/html');

		parent::display($tpl);
	}

	/**
	 * Add the page title and toolbar.
	 *
	 * @return  void
	 *
	 * @since   1.0.0
	 */
	protected function addToolbar()
	{
		$canDo = ContentHelper::getActions('com_installer');

		if ($canDo->get('core.edit.state'))
		{
			JToolbarHelper::publish('manage.publish', 'JTOOLBAR_ENABLE', true);
			JToolbarHelper::unpublish('manage.unpublish', 'JTOOLBAR_DISABLE', true);
			JToolbarHelper::divider();
		}

		JToolbarHelper::custom('manage.refresh', 'refresh', 'refresh', 'JTOOLBAR_REFRESH_CACHE');
		JToolbarHelper::divider();

		if ($canDo->get('core.delete'))
		{
			JToolbarHelper::deleteList('COM_INSTALLER_CONFIRM_UNINSTALL', 'manage.remove', 'JTOOLBAR_UNINSTALL');
			JToolbarHelper::divider();
		}

		JToolbarHelper::custom('extensions.export', 'share', 'share', 'PLG_SYSTEM_EXTENSIONEXPORT_BUTTON_EXPORT');

		JHtmlSidebar::setAction('index.php?option=com_installer&view=manage');

		parent::addToolbar();
		JToolbarHelper::help('JHELP_EXTENSIONS_EXTENSION_MANAGER_MANAGE');
	}
}


src/lib/Exporter.php
<?php
/**
 * @package     GreenCape Extension Exporter
 * @author      Niels Braczek <nbraczek@bsds.de>
 *
 * @copyright   Copyright (C) 2012 - 2017 BSDS Braczek Software- und DatenSysteme. All rights reserved.
 * @license     GNU General Public License version 3 or later; see LICENSE.txt
 */

namespace GreenCape\Extension;

use JFile;
use JFolder;
use Joomla\CMS\Factory;
use JPath;

jimport('joomla.application.module.controlleradmin');
jimport('joomla.filesystem.folder');
jimport('joomla.filesystem.file');

/**
 * @package     GreenCape\Extension
 *
 * @since       1.0.0
 */
class Exporter
{
	/**
	 * @var string
	 * @since 1.0.0
	 */
	private $filesTargetPath;

	/**
	 * @var string
	 * @since 1.0.0
	 */
	private $mediaSourcePath;

	/**
	 * @var string
	 * @since 1.0.0
	 */
	private $mediaTargetPath;

	/**
	 * @var string
	 * @since 1.0.0
	 */
	private $languageTargetPath;

	/**
	 * @var string
	 * @since 1.0.0
	 */
	private $adminFilesTargetPath;

	/**
	 * @var string
	 * @since 1.0.0
	 */
	private $adminLanguageTargetPath;

	/**
	 * The name of the package
	 *
	 * @var string
	 * @since 1.0.0
	 */
	private $packageName;

	/**
	 * The version of the extension
	 *
	 * @var string
	 * @since 1.0.0
	 */
	private $version;

	/**
	 * The name of the extension
	 *
	 * @var string
	 * @since 1.0.0
	 */
	private $extension;

	/**
	 * The type of the extension
	 *
	 * @var string
	 * @since 1.0.0
	 */
	private $type;

	/**
	 * The client ID
	 *
	 * 0 = administrator
	 * 1 = site
	 *
	 * @var int
	 * @since 1.0.0
	 */
	private $clientId;

	/**
	 * The group (type) of the plugin
	 *
	 * @var string
	 * @since 1.0.0
	 */
	private $pluginGroup;

	/**
	 * The working directory
	 *
	 * @var string
	 * @since 1.0.0
	 */
	private $exportDirectory;

	/**
	 * A subdirectory for collection of extension files
	 *
	 * @var string
	 * @since 1.0.0
	 */
	private $fileBucket;

	/**
	 * UNIX permissions for exported directories
	 *
	 * @var int
	 * @since 1.0.0
	 */
	private $dirMode;

	/**
	 * UNIX permissions for exported files
	 *
	 * @var int
	 * @since __DEPLOY_VERSION__
	 */
	private $fileMode;

	/**
	 * Exporter constructor.
	 *
	 * @param string $exportDirectory The working directory
	 * @param int    $dirMode         UNIX permissions for exported directories, defaults to 0755
	 * @param int    $fileMode        UNIX permissions for exported files, defaults to 0644
	 *
	 * @since 1.0.0
	 */
	public function __construct($exportDirectory, $dirMode = 0755, $fileMode = 0644)
	{
		$this->exportDirectory = $exportDirectory;
		$this->dirMode         = $dirMode;
		$this->fileMode        = $fileMode;
	}

	/**
	 * Export an extension
	 *
	 * @param string $extension   The extension
	 * @param string $type        The extension type
	 * @param int    $clientId    The client
	 * @param string $pluginGroup The plugin group
	 *
	 * @return string The package name
	 * @throws \Exception
	 * @throws \RuntimeException
	 * @throws \UnexpectedValueException
	 *
	 * @since 1.0.0
	 */
	public function export($extension, $type, $clientId, $pluginGroup)
	{
		$this->extension   = $extension;
		$this->fileBucket  = $extension;
		$this->type        = $type;
		$this->clientId    = $clientId;
		$this->pluginGroup = $pluginGroup;

		$manifest                      = $this->getManifest();
		$this->version                 = (string) $manifest->version;
		$this->filesTargetPath         = $this->getAttribute($manifest, 'files', 'folder', '');
		$this->mediaSourcePath         = $this->getAttribute($manifest, 'media', 'destination', $this->extension);
		$this->mediaTargetPath         = $this->getAttribute($manifest, 'media', 'folder', '');
		$this->languageTargetPath      = $this->getAttribute($manifest, 'languages', 'folder', '');
		$this->adminFilesTargetPath    = '';
		$this->adminLanguageTargetPath = '';

		if ($this->type === 'plugin')
		{
			$this->fileBucket = "plg_{$this->pluginGroup}_{$this->extension}";
		}

		if ($this->type === 'template')
		{
			$this->fileBucket = "tpl_{$this->extension}";
		}

		if (isset($manifest->administration))
		{
			$this->adminFilesTargetPath    = $this->getAttribute($manifest->administration, 'files', 'folder', '');
			$this->adminLanguageTargetPath = $this->getAttribute($manifest->administration, 'languages', 'folder', '');
		}

		$this->packageName = "{$this->fileBucket}-{$this->version}";

		if (JFolder::exists("{$this->exportDirectory}/{$this->fileBucket}"))
		{
			$this->removeArtifacts();
		}

		$this->copyExtension();

		if (isset($manifest->languages) || (isset($manifest->administration) && isset($manifest->administration->languages)))
		{
			$this->copyLanguages($manifest);
		}

		if (isset($manifest->media))
		{
			$this->copyMedia();
		}

		$this->zip();

		$this->fixPermissions();

		return $this->packageName;
	}

	/**
	 * Remove artifacts
	 *
	 * @since 1.0.0
	 */
	public function removeArtifacts()
	{
		JFolder::delete("{$this->exportDirectory}/{$this->fileBucket}");
	}

	/**
	 * Copy a media directory
	 *
	 * @throws \RuntimeException
	 *
	 * @since 1.0.0
	 */
	private function copyMedia()
	{
		$this->copyDirectory(
			JPATH_SITE . "/media/{$this->mediaSourcePath}",
			"{$this->exportDirectory}/{$this->fileBucket}/{$this->mediaTargetPath}"
		);
	}

	/**
	 * Copy a plugin directory
	 *
	 * @throws \RuntimeException
	 *
	 * @since 1.0.0
	 */
	private function copyPlugin()
	{
		$this->copyDirectory(
			JPATH_SITE . "/plugins/{$this->pluginGroup}/{$this->extension}",
			"{$this->exportDirectory}/{$this->fileBucket}"
		);
	}

	/**
	 * Copy a module directory
	 *
	 * @throws \RuntimeException
	 *
	 * @since 1.0.0
	 */
	private function copyModule()
	{
		$this->copyDirectory(
			($this->clientId === 0 ? JPATH_SITE : JPATH_ADMINISTRATOR) . "/modules/{$this->extension}",
			"{$this->exportDirectory}/{$this->fileBucket}"
		);
	}

	/**
	 * Copy a template directory
	 *
	 * @since 1.0.0
	 * @throws \RuntimeException
	 */
	private function copyTemplate()
	{
		$this->copyDirectory(
			($this->clientId === 0 ? JPATH_SITE : JPATH_ADMINISTRATOR) . "/templates/{$this->extension}",
			"{$this->exportDirectory}/{$this->fileBucket}"
		);
	}

	/**
	 * Copy a component
	 *
	 * @since 1.0.0
	 * @throws \RuntimeException
	 * @throws \Exception
	 */
	private function copyComponent()
	{
		$this->copyComponentAdministrator();
		$this->copyComponentSite();

		$baseName   = substr($this->extension, 4);
		$sourcePath = "{$this->exportDirectory}/{$this->fileBucket}/{$this->adminFilesTargetPath}/{$baseName}.xml'";

		if (!JFile::exists($sourcePath))
		{
			Factory::getApplication()->enqueueMessage("$sourcePath not found");
			$sourcePath = "{$this->exportDirectory}/{$this->fileBucket}/{$this->filesTargetPath}/{$baseName}.xml";
		}

		$targetPath = "{$this->exportDirectory}/{$this->fileBucket}/{$baseName}.xml";

		JFile::move($sourcePath, $targetPath);
	}

	/**
	 * Copy an extension
	 *
	 * @since 1.0.0
	 * @throws \RuntimeException
	 * @throws \Exception
	 */
	private function copyExtension()
	{
		switch ($this->type)
		{
			case 'component':
				$this->copyComponent();
				break;

			case 'module':
				$this->copyModule();
				break;

			case 'template':
				$this->copyTemplate();
				break;

			case 'plugin':
				$this->copyPlugin();
				break;

			default:
				throw new \RuntimeException('Unknown extension type ' . $this->type);
		}
	}

	/**
	 * Create a ZIP archive
	 *
	 * @since 1.0.0
	 * @throws \UnexpectedValueException
	 */
	private function zip()
	{
		$originalDirectory = getcwd();
		$workingDirectory  = "{$this->exportDirectory}/{$this->fileBucket}";
		chdir($workingDirectory);

		$zipper = new Zipper();

		foreach (JFolder::files('.', '', true, true) as $file)
		{
			$zipper->addFile(
				str_replace('./', '', $file),
				file_get_contents($file),
				filemtime($file)
			);
		}

		$zipper->create(JPath::clean("{$this->exportDirectory}/{$this->packageName}.zip"));

		chdir($originalDirectory);
	}

	/**
	 * Copy language files
	 *
	 * @param string            $language   The language code
	 * @param string            $sourcePath The source path
	 * @param string            $targetPath The target path
	 * @param \SimpleXMLElement $manifest
	 *
	 * @since 1.0.0
	 */
	private function copyLanguageFiles($language, $sourcePath, $targetPath, $manifest)
	{
		foreach (['ini', 'sys.ini'] as $ext)
		{
			$languageFile = "{$language}.{$this->fileBucket}.{$ext}";

			if (!JFile::exists("{$sourcePath}/{$languageFile}"))
			{
				continue;
			}

			$targetName = "{$language}/{$languageFile}";
			foreach ($manifest->language as $file)
			{
				if (basename($file) === $languageFile)
				{
					$targetName = $file;
					break;
				}
			}

			if (!empty($targetName))
			{
				JFolder::create(dirname("{$targetPath}/{$targetName}"), $this->dirMode);
				JFile::copy("{$sourcePath}/{$languageFile}", "{$targetPath}/{$targetName}");
			}
		}
	}

	/**
	 * Copy a language directory
	 *
	 * @param string            $sourcePath The source path
	 * @param string            $targetPath The target path
	 * @param \SimpleXMLElement $manifest
	 *
	 * @since 1.0.0
	 */
	private function copyLanguageDirectory($sourcePath, $targetPath, $manifest)
	{
		if (!JFolder::exists($sourcePath))
		{
			return;
		}

		foreach (JFolder::listFolderTree($sourcePath, $filter = '') as $folder)
		{
			$this->copyLanguageFiles($folder['name'], $folder['fullname'], $targetPath, $manifest);
		}
	}

	/**
	 * Copy language files if not in extension directory
	 *
	 * @param \SimpleXMLElement $manifest
	 *
	 * @since 1.0.0
	 */
	private function copyLanguages($manifest)
	{
		$locations = [
			'site'    => [
				'source'   => JPATH_SITE . '/language',
				'target'   => "{$this->fileBucket}/{$this->languageTargetPath}",
				'manifest' => $manifest->languages ?: null,
			],
			'admin'   => [
				'source'   => JPATH_ADMINISTRATOR . '/language',
				'target'   => "{$this->fileBucket}/{$this->adminLanguageTargetPath}",
				'manifest' => null,
			],
			'plugins' => [
				'source'   => JPATH_ADMINISTRATOR . '/language',
				'target'   => "{$this->fileBucket}/{$this->languageTargetPath}",
				'manifest' => null,
			],
		];

		if (isset($manifest->administration))
		{
			$locations['admin']['manifest'] = $manifest->administration->languages ?: null;
		}

		if ($this->type === 'plugin')
		{
			$locations['plugins']['manifest'] = $manifest->languages ?: null;
		}

		foreach ($locations as $location)
		{
			if ($location['manifest'] === null)
			{
				continue;
			}
			$this->copyLanguageDirectory(
				$location['source'], "{$this->exportDirectory}/{$location['target']}", $location['manifest']
			);
		}
	}

	/**
	 * Copy site parts of a component
	 *
	 * @since 1.0.0
	 * @throws \RuntimeException
	 */
	private function copyComponentSite()
	{
		$this->copyDirectory(
			JPATH_SITE . "/components/{$this->extension}",
			"{$this->exportDirectory}/{$this->fileBucket}/{$this->filesTargetPath}"
		);
	}

	/**
	 * Copy admin parts of a component
	 *
	 * @since 1.0.0
	 * @throws \RuntimeException
	 */
	private function copyComponentAdministrator()
	{
		$this->copyDirectory(
			JPATH_ADMINISTRATOR . "/components/{$this->extension}",
			"{$this->exportDirectory}/{$this->fileBucket}/{$this->adminFilesTargetPath}"
		);
	}

	/**
	 * Copy a directory
	 *
	 * @param $sourcePath
	 * @param $targetPath
	 *
	 * @throws \RuntimeException
	 *
	 * @since 1.0.0
	 */
	private function copyDirectory($sourcePath, $targetPath)
	{
		if (!JFolder::exists($sourcePath))
		{
			return;
		}

		JFolder::create($targetPath, $this->dirMode);
		JFolder::copy($sourcePath, $targetPath, '', true);
	}

	/**
	 * Get the manifest
	 *
	 * @return \SimpleXMLElement
	 *
	 * @since 1.0.0
	 * @throws \RuntimeException
	 */
	private function getManifest()
	{
		return simplexml_load_string(file_get_contents($this->getManifestPath()));
	}

	/**
	 * Get the manifest path
	 *
	 * @return string
	 *
	 * @since 1.0.0
	 * @throws \RuntimeException
	 */
	private function getManifestPath()
	{
		$clientPath   = $this->clientId === 0 ? JPATH_SITE : JPATH_ADMINISTRATOR;
		$manifestName = substr($this->extension, 4) . '.xml';
		$manifestPath = [
			'component' => "{$clientPath}/components/{$this->extension}/{$manifestName}",
			'module'    => "{$clientPath}/modules/{$this->extension}/mod_{$manifestName}",
			'plugin'    => JPATH_SITE . "/plugins/{$this->pluginGroup}/{$this->extension}/{$this->extension}.xml",
			'template'  => "{$clientPath}/templates/{$this->extension}/templateDetails.xml",
		];

		if (!isset($manifestPath[$this->type]))
		{
			throw new \RuntimeException("Extensions of type '{$this->type}' are not supported.");
		}

		if (!JFile::exists($manifestPath[$this->type]))
		{
			throw new \RuntimeException("No manifest found for {$this->type} {$this->extension} (expected {$manifestPath[$this->type]})");
		}

		return $manifestPath[$this->type];
	}

	/**
	 * Safely get an attribute (or its default) from a SimpleXMLElement
	 *
	 * @param \SimpleXMLElement $element
	 * @param string            $tag
	 * @param string            $attribute
	 * @param string            $default
	 *
	 * @return string
	 *
	 * @since 1.0.0
	 */
	private function getAttribute($element, $tag, $attribute, $default)
	{
		if (!isset($element->{$tag}))
		{
			return (string) $default;
		}

		return (string) $element->{$tag}->attributes()->{$attribute} ?: $default;
	}

	/**
	 * Fix permissions
	 *
	 * @since __DEPLOY_VERSION__
	 */
	private function fixPermissions()
	{
		foreach (JFolder::files("{$this->exportDirectory}/{$this->fileBucket}", '.', 10, true) as $file)
		{
			chmod($file, $this->fileMode);
		}

		foreach (JFolder::folders("{$this->exportDirectory}/{$this->fileBucket}", '.', 10, true) as $folder)
		{
			chmod($folder, $this->dirMode);
		}

		chmod("{$this->exportDirectory}/{$this->packageName}.zip", $this->fileMode);
	}
}
